lib/optimizely/config_manager/polling_config_manager.rb

```diff
@@ -0,0 +1,156 @@
+#    limitations under the License.
+#
+require_relative '../error_handler'
+require_relative 'base_config_manager'
+require_relative 'exceptions'
+require_relative 'helpers/constants'
+require_relative 'helpers/validator'
+require_relative 'logger'
+require_relative 'notification_center'
+require_relative 'project_config'
+require_relative 'static_config_manager'
+
+require 'net/http'
+
+module Optimizely
+  class PollingConfigManager < StaticConfigManager
+    # Config manager that polls for the datafile and updated ProjectConfig based on an update interval.
+
+    # Initialize config manager. One of sdk_key or url has to be set to be able to use.
+    #
+    # sdk_key - Optional string uniquely identifying the datafile.
+    # datafile: Optional JSON string representing the project.
+    # update_interval - Optional floating point number representing time interval in seconds
+    #                  at which to request datafile and set ProjectConfig.
+    # url - Optional string representing URL from where to fetch the datafile. If set it supersedes the sdk_key.
+    # url_template - Optional string template which in conjunction with sdk_key
+    #               determines URL from where to fetch the datafile.
+    # logger - Provides a logger instance.
+    # error_handler - Provides a handle_error method to handle exceptions.
+    # skip_json_validation - Optional boolean param which allows skipping JSON schema
+    #                       validation upon object invocation. By default JSON schema validation will be performed.
+    def initialize(
+      sdk_key = nil,
+      datafile = nil,
+      update_interval = nil,
+      url = nil,
+      url_template = nil,
+      logger = nil,
+      error_handler = nil,
+      skip_json_validation = false
+    )
+
+      super(logger, error_handler)
+      url_template ||= Helpers::Constants::CONFIG_MANAGER['DATAFILE_URL_TEMPLATE']
+      @datafile_url = get_datafile_url(sdk_key, url, url_template)
+      @update_interval = update_interval || Helpers::Constants::CONFIG_MANAGER['DEFAULT_UPDATE_INTERVAL']
+      update_interval(update_interval)
+      @last_modified = nil
+      @config = nil
+      @validate_schema = !skip_json_validation
+      @polling_thread = Thread.new { run }
+      set_config(datafile) if datafile
+    end
+
+    def get_datafile_url(sdk_key, url, url_template)
+      # Helper method to determine URL from where to fetch the datafile.
+      # sdk_key - Key uniquely identifying the datafile.
+      # url - String representing URL from which to fetch the datafile.
+      # url_template - String representing template which is filled in with
+      #               SDK key to determine URL from which to fetch the datafile.
+      # Returns String representing URL to fetch datafile from.
+
+      raise InvalidInputsError, 'Must provide at least one of sdk_key or url.' if sdk_key.nil? && url.nil?
+
+      return (url_template % sdk_key) unless url
```

Comment:
can this raise an exception if url_template doesn't have a placeholder or more? or the type is different? If yes, we should catch it